flink-ml-iteration/flink-ml-iteration-common/src/main/java/org/apache/flink/iteration/operator/feedback/SpillableFeedbackQueue.java

```diff
@@ -0,0 +1,115 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.iteration.operator.feedback;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.api.common.typeutils.TypeSerializer;
+import org.apache.flink.core.memory.DataInputView;
+import org.apache.flink.core.memory.DataOutputSerializer;
+import org.apache.flink.core.memory.MemorySegment;
+import org.apache.flink.runtime.io.disk.InputViewIterator;
+import org.apache.flink.runtime.io.disk.SpillingBuffer;
+import org.apache.flink.runtime.io.disk.iomanager.IOManager;
+import org.apache.flink.runtime.memory.MemoryAllocationException;
+import org.apache.flink.runtime.memory.MemoryManager;
+import org.apache.flink.table.runtime.operators.sort.ListMemorySegmentPool;
+import org.apache.flink.util.MutableObjectIterator;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * * A queue that can spill the items to disks automatically when the memory buffer is full.
+ *
+ * @param <T> The element type.
+ */
+@Internal
+final class SpillableFeedbackQueue<T> {
+    private final DataOutputSerializer output = new DataOutputSerializer(256);
+    private final TypeSerializer<T> serializer;
+    private final IOManager ioManager;
+    private final MemoryManager memoryManager;
+    private final int numPages;
+
+    private List<MemorySegment> segments;
+    private ListMemorySegmentPool segmentPool;
+
+    private SpillingBuffer target;
+    private long size = 0L;
+
+    SpillableFeedbackQueue(
+            IOManager ioManager,
+            MemoryManager memoryManager,
+            TypeSerializer<T> serializer,
+            long inMemoryBufferSize,
+            long pageSize)
+            throws MemoryAllocationException {
+        this.serializer = Objects.requireNonNull(serializer);
+        this.ioManager = Objects.requireNonNull(ioManager);
+        this.memoryManager = Objects.requireNonNull(memoryManager);
+
+        this.numPages = (int) (inMemoryBufferSize / pageSize);
+        resetSpillingBuffer();
+    }
+
+    void add(T item) {
+        try {
+            output.clear();
+            serializer.serialize(item, output);
+            target.write(output.getSharedBuffer(), 0, output.length());
+            size++;
+        } catch (IOException e) {
+            throw new IllegalStateException(e);
+        }
+    }
+
+    MutableObjectIterator<T> iterate() {
+        try {
+            DataInputView input = target.flip();
+            return new InputViewIterator<>(input, this.serializer);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    long size() {
+        return size;
+    }
+
+    void reset() throws Exception {
+        size = 0;
+        close();
```

Comment:
close() is generally used to release all resources used by this instance and there should be no further invocation of any method of this instance after close() is invoked.

It would be useful to follow this pattern.